docs/DEPLOYMENT.md

```diff
@@ -0,0 +1,141 @@
+```
+yarn run deployStaging
+```
+
+These scripts will first build the game code and the electron application, after which the image will be created based on the `Dockerfile`. Built game code and the electron application will be copied over to our final image, which will then be uploaded to the Docker hub. Since our image will be run on a Linux host machine, we also had to build the image using Linux containers, which Docker on Windows helpfully allows us to do.
+
+The name of the built image will be `shieldbattery/SERVER_NAME:VERSION`, where `SERVER_NAME` is either `production` or `staging` and `VERSION` is whatever's written in `package.json`'s `version` field. However, every time a new version is deployed, a `shieldbattery/SERVER_NAME:latest` tag will be updated to point to the new image.
+
+NOTE: The deploy script doesn't particularly care which branch you're trying to deploy. It will just take whatever code is currently in the repository and use that.
+
+## Running the server image inside a Docker container
+
+This is the second (and final) step in our deployment process. This step should be done on a Linux operating system, generally on the publicly available machine that's used to host the servers.
+
+### Create config files
+
+ShieldBattery's Docker image requires two config files to be created, which will then be mounted into it once the container starts. Those files are `config.js` and `database.json`, and you can find an example of them in the repository, inside the `server` folder. The location of the files doesn't generally matter when running the application through Docker, as their path can be configured through environment variables, as explained in the below section.
+
+Important thing to note when editing those files is that for the `host` part in redis and postgres connection strings, their service name should be used as defined in the `docker-compose.yml` file, which is `redis` and `db` respectively. For example, a postgres connection string would be: `postgres://USER:PASSWORD@db/shieldbattery`.
+
+### Setup environment variables
+
+Docker allows for a couple of ways in which the `docker-compose.yml` file can be configured when running the containers, and all of them are centered around environment variables. In short, they are:
+- The `.env` file - a single file which can hold all of the environment variables you need; needs to be in the same directory as the `docker-compose.yml` file.
+- Shell environment variables - you can set the environemnt variables directly in shell before running the docker compose (they take priority over the `.env` file).
+
+You can read more about the environment variables in docker compose at [this link](https://docs.docker.com/compose/environment-variables/).
+
+The following environment variables are available to be configured (those without a default value are required):
+- `SERVER_NAME` - Must be either `production` or `staging` which is used to fetch the image from docker hub.
```

Comment:
Maybe I'm missing something, but I don't really understand the purpose of this. Generally the staging server should be running the same stuff as the production one, it's just a later version (before that version gets promoted to production). If we have to change things other than configuration between staging and production then staging isn't a particularly good test. Is there something I'm overlooking there?

Reply:
I assume from doing more reading that this is purely for the electron app side of things. I wonder if it might be better to upload versions the same package for staging/production server URLs. From a quick look at Docker Hub it seems like these versions can include strings and the like: https://hub.docker.com/_/sonarqube/?tab=tags

So we could upload like a 1.0.0-production and 1.0.0-staging at the same time. This still feels quite odd to me though given that both images would now be including a URL to our actual servers, rather than being deployable to any random location. If we instead mounted the installer location (or made a separate service for delivering the installers) perhaps all that weirdness goes away. That'd also remove all the stuff preventing us from easily building these docker images remotely.

Reply:
I can't remember my exact motivation now, and this is something I did very early in my process of learning Docker, but I _think_ my idea of this environment variable was to use it as just a decorator in the image name; I don't think it's used in any more meaningful capacity than that. For some reason, I thought back then that we needed different images for different flavors of the server.

But thinking about it again now, I see what you mean that this doesn't make much sense. The (only?) difference between our production and staging servers is located in the config file, which is loaded externally into the image, so I think you're right, we could probably deploy only one image.

I think all notions of deploying a particular server can be removed, except one: running `dist` script before deploying the production server, and running `diststaging` before deploying the staging server. So we probably still need two npm scripts for deploying, but we don't need to send the `server` argument anymore, just the version here:
https://github.com/ShieldBattery/ShieldBattery/pull/414/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R20-R21

Reply:
Took me a while, but I actually just understood what you mean by mounting the installer location. And yeah, I guess the only thing holding us back from having server-agnostic deployment is the way we build electron stuff with `electron-builder`. If we can somehow consolidate `prod.yml` and `staging.yml` into one build file, then deployment, along with building, won't depend on the actual server.